components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java

```diff
@@ -282,4 +294,79 @@ private void removeEntriesFromCache(ServiceProvider serviceProvider, String tena
+        Object saasStatus = IdentityUtil.threadLocalProperties.get().get(SAAS_PROPERTY);
+        if (saasStatus != null) {
+            saasEnabled = (Boolean) saasStatus;
+        }
+        if ((saasEnabled == true) && (serviceProvider.isSaasApp() == false)) {
+            if (log.isDebugEnabled()) {
+                log.debug("SaaS is disabled.");
+            }
+            InboundAuthenticationRequestConfig[] configs = serviceProvider.getInboundAuthenticationConfig()
+                    .getInboundAuthenticationRequestConfigs();
+            try {
+                for (InboundAuthenticationRequestConfig config : configs) {
+
+                    if (IdentityApplicationConstants.OAuth2.NAME.equalsIgnoreCase(config.getInboundAuthType()) &&
+                            config.getInboundAuthKey() != null) {
+                        String oauthKey = config.getInboundAuthKey();
+                        Set<String> accessTokens = tokenMgtDAO.getActiveTokensForConsumerKey(oauthKey);
+                        for (String accessToken : accessTokens) {
+                            AccessTokenDO accessTokenDO = null;
+                            accessTokenDO = tokenMgtDAO.retrieveAccessToken(accessToken, false);
+                            OAuthConsumerDAO oAuthConsumerDAO = new OAuthConsumerDAO();
+                            String clientSecret = oAuthConsumerDAO.getOAuthConsumerSecret(accessTokenDO.getConsumerKey());
+                            String userTenantDomain = accessTokenDO.getAuthzUser().getTenantDomain();
+
+                            if (!StringUtils.equals(tenantDomain, userTenantDomain)) {
+                                if (log.isDebugEnabled()) {
+                                    log.debug("user tenant domain does not match with application tenant domain.");
+                                }
+                                OAuthRevocationRequestDTO revokeReqDTO = new OAuthRevocationRequestDTO();
+                                revokeReqDTO.setConsumerKey(accessTokenDO.getConsumerKey());
+                                revokeReqDTO.setConsumerSecret(clientSecret);
+                                revokeReqDTO.setToken(accessTokenDO.getAccessToken());
+                                revokeReqDTO.setToken_type(accessTokenDO.getTokenType());
+                                OAuthRevocationResponseDTO oAuthRevocationResponseDTO = OAuth2ServiceComponentHolder.getOAuth2Service().revokeTokenByOAuthClient(revokeReqDTO);
+
+                                if (!oAuthRevocationResponseDTO.isError()) {
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("Access token revoked for client ID: " + config.getInboundAuthKey());
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            } catch (IdentityException ex) {
```

Comment:
Shall we have the catch blocks inside the for loops as much as possible. So that failure to revoke one will not prevent the rest being revoked.